Convert from pysis to kalasiris and rewrite of spatial/isis.py

rbeyer opened the pull request:
* Converted everything I could find from *pysis* to *kalasiris*.
* Added tests for functions that used *kalasiris* in order to make sure I didn't screw anything up (fingers crossed).
* Added a Makefile with all kinds of functionality, but  specifically a "test-resources" target to download a pair of sample HiRISE files to run tests on.  If this directory isn't present, the tests that need them will be safely skipped.
* Reviewed the three functions in spatial/isis.py, and refactored them.
  * They were inconsistently returning single values or a list, now, if given a single value, they'll return a two-tuple of singletons, if they are given a Sequence or a Numpy array, they'll return a two-tuple of Sequences or Numpy arrays, as appropriate.
  * Coordinate order shall be: x then y, longitude then latitude, sample then line, anything else is confusing and wrong.  If thou takest arguments, thou shall takest them in that order.  If thou returnest values or array-likes in a tuple, thou shall returnest them in that order. Functions refactored to support that consistency, and also refactored functions that call these to also work under these new commandments.  The rest of the library probably needs to be scrubbed.

This is a lot to review, I know.  Sorry-not-sorry.
 
This would close #570

rbeyer commented:
Needs ISIS in the test runner.  Not sure how you want to approach it?  With only mock tests running, I think you could get away with just setting ISISROOT and ISISDATA to dummy values.  If you wanted to actually test the things that run ISIS, then you'll have to install a real ISIS and a real ISISDATA, I guess.

rbeyer commented:
Still skipping the tests on the CI, not entirely sure why, probably has to do with paths on the CI runner, but at least everything completes, I guess.

jlaura commented on autocnet/spatial/isis.py:
```diff
@@ -1,167 +1,260 @@
 
-    if isinstance(x, Number) and isinstance(y, Number):
-        x, y = [x], [y]
+    if isinstance(x, abc.Sequence) and isinstance(y, abc.Sequence):
```
So this change is in order to support a sequence of numbers? Right now, couldn't a user pass in any arbitrary iterable (since they have the magic methods for a sequence) and this check would pass? I guess, my question is, why not check that it is not a string and then check the length?

This solution will work, I'm just curious about why abc.Sequence.

rbeyer replied:
Sure.  This was a fun one, and I may have screwed it up.

Can they pass in an arbitrary iterable?  No, an "iterable" would fail this isinstance test.  But I know that's not what you mean, you mean an object that can be iterated on.  That's right.  So why not: Python sets are not strings, have a length, but cannot be iterated on in the context that this function needs, so would be bad.  In general, testing against abc's usually provides the "right" test (and also makes you think about how that object is going to be used in your function).

jlaura commented on autocnet/spatial/isis.py:
```diff
@@ -1,167 +1,260 @@
+    if isinstance(x, (abc.Sequence, np.ndarray)):
+        return results
+    else:
+        return results[0]
```
I appreciate encapsulating the return length logic in here so that the caller get a response identical to the input argument length. :+1:

rbeyer replied:
Yeah, I went back and forth about whether to implement that as two user-facing functions (so you had a function that took single numbers, and one that took many), but this is kind of what numpy does, so I figured that was as good a model as any.

jlaura reviewed (approved):
@rbeyer Thanks for the submission! Minor comments inline. 

@Kelvinrr Can you also give this a look. We should figure out if/how to get the change detection stuff into a consuming library most likely or discuss/accept that this should be a mono-repo and discuss how to test/deploy autocnet for control with autocnet for change detection.

Kelvinrr commented on autocnet/matcher/ground.py:
```diff
@@ -163,7 +163,7 @@ def propagate_ground_point(point,
         if cam_type == "isis":
             try:
-                line, sample = isis.ground_to_image(image["path"], lon_oc, lat_oc)
+                sample, line = isis.ground_to_image(image["path"], lon_oc, lat_oc)
```
Yeah, I am almost thinking of having this in a non-dev branch to test it first to make sure nothing has broken. And if we choose to make a release, we got to do a major version tick? I agree making them all the same, since it's a huge pain, but it's also a problem with the libraries we use.

rbeyer replied:
By all means.  I think I made all the conversions so that the things that we could control in this library (that touched return values from these functions) were in that order.

rbeyer commented on Makefile:
```diff
@@ -0,0 +1,99 @@
+.PHONY: clean clean-test clean-pyc clean-build docs help
```
No, just copied and modified from our PlanetaryPy package template: https://github.com/planetarypy/package-template/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/Makefile

Surely there are a million different ways to realize test artifacts and such.  I'm not hard over on this, it was just the easiest thing for me to grab and use.

rbeyer commented:
There is also *at least* something broken with the `graph.network.NetworkCandidateGraph.to_isis()` function when I use this new branch, but I cannot for the life of me figure out what it is.  I basically get this warning:
```
warnings.warn(f'{f} in candidate graph but not in output network.')
```
I have this simple network with only four images, and it gives this warning for two of them, but writes out the other two, so it isn't completely broken, just *kinda* broken.  Maybe when the `io_controlnetwork.db_to_df()` happens, it pulls data from the DB that was put in differently than is being expected when `df['imageid'].unique()` gets called?  Maybe there’s something else in `self.nodes[i]['data']['image_path’]` than is expected?  Not entirely sure.  I tried to trace it a little, but quickly got lost in the labyrinth of the extremely complex NetworkCandidateGraph class.

jlaura commented:
> There is also _at least_ something broken with the `graph.network.NetworkCandidateGraph.to_isis()` function when I use this new branch, but I cannot for the life of me figure out what it is. I basically get this warning:
> 
> ```
> warnings.warn(f'{f} in candidate graph but not in output network.')
> ```

I think that this is related to #545. I have some time this fine morning, so I'll have a look at that issue to see if I can't get it resolved. Otherwise, I think the thing for me to do with this PR is to pull it into a branch and run some end-to-end tests, mainly looking to ensure that the line, sample updates are all good and to see if we have any weird ISIS issues. I don't expect either, but I want to be really cautious.

Kelvinrr reviewed (approved):
Just some comments inline, still good as is.

rbeyer commented on autocnet/spatial/isis.py:
```diff
@@ -1,167 +1,260 @@
 
 
-def point_info(cube_path, x, y, point_type, allow_outside=False):
+def point_info(
+        cube_path: os.PathLike,
+        x,
+        y,
+        point_type: str,
+        allowoutside=False
+):
```
Sure, there didn't seem to be a style guide in this code, and I just reflexively follow the black style, although I didn't run black on these files, because I was trying to keep the diffs minimal.

rbeyer commented on autocnet/spatial/isis.py:
```diff
@@ -1,167 +1,260 @@
-    if isinstance(x, Number) and isinstance(y, Number):
-        x, y = [x], [y]
+    if isinstance(x, abc.Sequence) and isinstance(y, abc.Sequence):
+        if len(x) != len(y):
+            raise IndexError(
+                f"Sequences given to x and y must be of the same length."
+            )
+        x_coords = x
+        y_coords = y
+    elif isinstance(x, Number) and isinstance(y, Number):
+        x_coords = [x, ]
+        y_coords = [y, ]
+    elif isinstance(x, np.ndarray) and isinstance(y, np.ndarray):
+        if not all((x.ndim == 1, y.ndim == 1)):
+            raise IndexError(
+                f"If they are numpy arrays, x and y must be one-dimensional, "
+                f"they were: {x.ndim} and {y.ndim}"
+            )
+        if x.shape != y.shape:
+            raise IndexError(
+                f"Numpy arrays given to x and y must be of the same shape."
+            )
+        x_coords = x
+        y_coords = y
+    else:
+        raise TypeError(
+            f"The values of x and y were neither Sequences nor individual "
+            f"numbers, they were: {x} and {y}"
+        )
```
Yeah, I'll admit this seemed "heavy" and there may be a more elegant way to express it.  However, that's independent of what you're talking about, and yes, if this is a repeated pattern in other functions, sure.

rbeyer commented on autocnet/spatial/isis.py:
```diff
@@ -1,167 +1,260 @@
+            ]
+
+        # ISIS's campt needs points in a file
+        with isis.fromlist.temp(p_list) as f:
+            cp = isis.campt(
+                cube_path,
+                coordlist=f,
+                allowoutside=allowoutside,
+                usecoordlist=True,
+                coordtype=point_type
+            )
```
Thanks!  I taught myself how to write Python context managers when I was implementing the kalasiris.fromlist module, and I'm not gonna lie, I was pretty proud of myself.

rbeyer commented:
> There is also _at least_ something broken with the `graph.network.NetworkCandidateGraph.to_isis()` function when I use this new branch, but I cannot for the life of me figure out what it is. ...

I had a moment to work this.  Jay was right: if I use the custom SQL that Lauren details in #545, then all four of my little test images end up in the .lis file, as do all of the measures and points.

Just to summarize, I have 4 images, which have 83 points and 197 measures after a call to `place_points_in_overlap()`.
If I run the default `ncg.to_isis()` with the current dev branch (with *pysis*), I get all four images in the output, 20 points and 40 measures (which presumably represents some sort of culling).

If I switch to this `kalasiris-570` branch, and do the same, when I run `ncg.to_isis()` I get the above-mentioned warnings, only get two images in the output, and 4 points and 7 measures in the .net file.

If I then re-run just the `ncg.to_isis()` but using Lauren's SQL, I get no warnings, all four images in my output with 83 points and 197 measures.

So I'm still not sure why switching from *pysis* to *kalasiris* has such drastic changes in the output of `to_isis()`.  Seems like it should be the same.  However, I suppose if the logic is going to undergo some change, then maybe this difference isn't important?

rbeyer commented:
Nope, I broke something in this PR, for sure (still the SQL change is good to know about).  Running `subpixel_register_point()` with `verbose=True` is very illustrative and shows that the matches coming out of this PR branch are terrible, and the dev branch matches are substantially better.  So I screwed something up somewhere and will try and see if I can find out where. :)

rbeyer commented:
Aha!  Fixed it.  Turns out I had misinterpreted the `[::-1]` on the tail of some lines in subpixel.py that *did* switch the output of the old functions to sample/line order, but I didn't realize it so I also reversed them.  Fixed that, and now this PR branch's output matches that of the default dev branch.

Kelvinrr commented:
@jlaura Did the recent change fix the problem? If so, can this get merged?

jlaura reviewed (approved):
The only thing this needs is a version number bump (0.5 to 0.5.1? - this is an internal API change). That change need to be in the setup.py file and perhaps in the recipe meta.yml file.

jlaura commented:
@Kelvinrr Have you been using this in your work this week?

Kelvinrr commented:
@jlaura I haven't since I was trying to get things working with your changes merged with dev and this was not merged yet.
